articles/110_interface_definition.md

```diff
@@ -0,0 +1,214 @@
+
+
+### Field types
+
+The type of a field can be either a primitive type or another data structure.
+Each of these can optionally be a dynamically or statically sized array.
+
+
+#### Primitive field types
+
+The following primitive types are defined:
+
+* `bool`
+* `byte`
+* `char`
+* `float32`, `float64`
+* `int8`, `uint8`
+* `int16`, `uint16`
+* `int32`, `uint32`
+* `int64`, `uint64`
+* `string`
+
+<div class="alert alert-warning" markdown="1">
+  <b>TODO:</b> consider <code>char</code>, <code>wstring</code>, <code>u16string</code>, <code>u32string</code>
+</div>
+
+
+#### Non-primitive field types
+
+Beside the primitive types other messages can be referenced to describe the type of a "complex" field.
+A complex field type is identified by a package and a message name.
+
+
+#### Arrays with static size
+
+A static array has exactly `N` elements of the specified type.
+`N` must be greater than `0`.
+
+
+#### Arrays with dynamic size
+
+A dynamic array can have between `0` and `N` elements of the specified type.
+`N` might not have an upper bound and may only be limited by the memory or other system specific limitations.
+
+
+#### Upper boundaries
+
+The size of *strings* as well as *dynamic arrays* can be limited with an *upper boundary*.
+This enables the preallocation of memory for data structures which use dynamically sized data.
+
+
+### Default values
+
+A field can optionally specify a default value.
+If no default value is specified a common default value is used:
+
+* for `bool` it is `false`
+* for *numeric* types it is a `0` value
+* for `string` it is an *empty* string
+
+<div class="alert alert-warning" markdown="1">
+  <b>TODO:</b> default values are currently not supported for <code>string array</code> fields and <i>complex</i> fields
+</div>
+
+
+### Constants
+
+Constants are defined by a *primitive type*, a *name* as well as a *fixed value*.
+
+
+## Conventions
+
+### Naming of messages and services
+
+Each file contains a single message or service.
+Message files use the extension `.msg`, service files use the extension `.srv`.
+
+Both file names must use an upper camel case name and only consist of alphanumeric characters.
+
+
+### Naming of fields
+
+Field names must be lowercase alphanumeric characters with underscores for separating words.
+They must not start or end with an underscore and never have two consecutive underscores.
+
+
+### Naming of constants
+
+Constant names must be uppercase alphanumeric characters with underscores for separating words.
+They must not start or end with an underscore and never have two consecutive underscores.
+
+
+## Syntax
+
+The message and service definitions are text files.
+
+
+### Comments
+
+The character `#` starts a comment, which terminates at the end of the line on which it occurs.
+
+
+### Message file format
+
+A line can either a field definition or a constant definition.
+
+
+#### Field definition
+
+A field definition has the following structure:
+
```

Comment:
I tried to describe that in e87d0005bb93a0c4c7379c0d1bbd54663f9892b4

But ultimately we should provide a BNF for it to avoid any interpretations.